### DISCUSSION TITLE
Adding new tutorial: starting from step file, meshing with pygmsh, and then, solving with fenics.

### DISCUSSION THREAD

**Discussion by fghoussen** [jorgensd/dolfinx-tutorial#134] [Ideas]
I suggest here an idea for adding a new tutorial: starting from step file, meshing with pygmsh, and then, solving with fenics.  

I succeeded to run chapter2/linearelasticity_code.py which has automatic mesh and fenics solve.

I succeeded to run https://gitlab.onelab.info/gmsh/gmsh/-/blob/master/examples/api/step_assembly.py which has custom mesh from CAD (STEP file) with physical surfaces (where you can apply BC on them), but, no fenics solve.

I didn't succeed to merge both tutorials. As this is a common use case, I guess having a new tutorial on this topic would be worth! That is:

- From a minimal STEP file (a 3D rod with 2 physical surfaces at both ends of the rod)
- Mesh it with `pygmsh` (3D mesh + export physical surfaces in the .msh file)
- With fenics, apply BC: clamp and traction force at both ends of the rod (using physical surfaces)
- Solve with fenics and visualize the elongated rod

I didn't succeed to design this tutorial which was the one I needed in the first place... But I would be glad somebody could succeed where I failed! :)

**Comment by jorgensd** [comment]
Could you have a look at: https://jsdokken.com/dolfinx-tutorial/chapter3/subdomains.html#subdomains-defined-from-external-mesh-data
As I believe this tutorial solves 98% of the functionality above

**Comment by fghoussen** [comment]
Thanks, missed this one. At work now, I had a bird-eye view. I'll have a closer look at the week-end.
Seems close but in 2D and maybe still a bit challenging for me (get physical surf from step file, apply BC on physical surf) as I am new to gmsh/fenics.

**Reply by jorgensd** [reply]
2D or 3D doesn't matter from a DOLFINx perspective. The point of the tutorial is that you can either read the mesh and tags in with `dolfin.Io.gmshio.read_from_msh` or create an XDMFFile with meshio and use XDMFFile.read/XDMFFile.read_meshtags to read in the data.

Ive Also got tutorials on pygmsh/Gmsh at 

https://jsdokken.com/src/tutorial_gmsh.html

https://jsdokken.com/src/pygmsh_tutorial.html
